### DISCUSSION TITLE
[gtk:x64-windows] Extra directory depth required for GTK4

### DISCUSSION THREAD

**Discussion by weirdworkshop** [microsoft/vcpkg#20672] [Q&A]
**Describe the bug**

when I try to #include the GTK-4.0 library the files  are located  in a extra directory so rather than   #include<gtk/gtk.h>  Visual Studio can see them in #include <gtk-4.0/gtk/gtk.h>  but the files make no reference to this extra directory depth.    See attached screen shots. 
<img width="666" alt="gtk_1" src="https://user-images.githubusercontent.com/32969686/124179670-11555580-daab-11eb-9f35-f6161ec254a7.png">
<img width="559" alt="gtk_2" src="https://user-images.githubusercontent.com/32969686/124179672-11edec00-daab-11eb-97fb-ebc4aa7c0531.png">
<img width="594" alt="gtk_3" src="https://user-images.githubusercontent.com/32969686/124179674-11edec00-daab-11eb-9c1a-d6bb23987e6a.png">

This also affect the other libraries that are installed as part of GTK4 such as Glib-2.0 

So have had to manually add in the extra directories into all the #include's   in all  the .h files 

**Environment**
- OS:  Windows 10 Home 21H1  Build 19043.1081
- Compiler: Visual Studio 19 Community Edition Version 16.10.3 

**To Reproduce**

using the Visual Studio x64 Native Tools Command Prompt

git clone https://github.com/Microsoft/vcpkg.git
.\vcpkg\bootstrap-vcpkg.bat
 .\vcpkg install gtk:x64-windows
 .\vcpkg integrate install

Created a new project in Visual Studio. 
Changed the file extension to .c
Changed the solution platform to x64
Copied and pasted the hello world program from https://www.gtk.org/docs/getting-started/hello-world/

Compiler cannot find gtk.h also indicated by Squiggly red lines from intellisense and does not compile.

After going into I think all of the .h files and adding the extra directory names into the #include statements the files are now found and the program compiles and runs.

**Expected behaviour**
The gtk.h and all library files are found, the hello world  program compiles and runs. 

**Failure logs** 
-(please attached failure logs)

**Additional context**

**Comment by NancyLi1013** [comment]
Hi @weirdworkshop 

Thanks for posting this issue.

Since the headers for `gtk ` were installed in _include/gtk-4.0/_ directory. But vcpkg can only find the headers in _include_ directory by default. In order to avoid conflicts, vcpkg add additional directory for these ports. In this case, we need to add ..\include\xxx\ path to msbuild project manually.

So for `gtk`, you need to add _VCPKG_ROOT\installed\x64-windows\include\gtk-4.0_ to your project, of course, its dependencies are also needed to add if those ports have the same case.

**Comment by yonaka** [comment]
I've been believing vcpkg does not require VS setting to use a library...
Even if it is unavoidable, the include path should not refer to the version number because the package name "gtk" does not refer to it.

**Comment by NancyLi1013** [comment]
> I've been believing vcpkg does not require VS setting to use a library...
> Even if it is unavoidable, the include path should not refer to the version number because the package name "gtk" does not refer to it.

Thanks for your focus on this issue. As I said above, you need to add the path to your project manually, then include the headers as normal.

**Comment by Hemaolle** [comment]
> I've been believing vcpkg does not require VS setting to use a library...
> Even if it is unavoidable, the include path should not refer to the version number because the package name "gtk" does not refer to it.

Yeah, at least the vcpkg readme promises:

> After this, you can now create a New non-CMake Project (or open an existing one). All installed libraries are immediately ready to be #include'd and used in your project without additional configuration.

**Comment by mihai-vlc** [comment]
Writing this here in case part of these steps might help someone in the future.

I attempted to run this over the weekend as an exercise on how to use vcpkg with visual studio and an external library and somehow I picked gtk.
I am not a C++ dev, just trying this out to see how difficult it is.

In the console I ran the following:
```
vcpkg install gtk  # this took a long time to complete ~2h for some reason, I also needed to run this from pwsh.exe, from cmd it failed saying that it can't find powershell core even though it was installed from the MS Store
vcpkg integrate project
```

I created a visual studio c++ empty project and added the following `vcpkg.json`
```
{
  "name": "example",
  "version-string": "0.0.1",
  "dependencies": [
    "gtk"
  ]
}
```

With a project open, go to Tools->NuGet Package Manager->Package Manager Console and paste:
    Install-Package vcpkg.C.tools.vcpkg -Source "C:\tools\vcpkg\scripts\buildsystems"

I went to Project Properties > vcpkg
Use Vcpkg = yes
Use Vcpkg Manifest = yes

I created a main.cpp file with the following (taken from the gtk homepage)
```
// Include gtk
#include <gtk/gtk.h>

static void on_activate (GtkApplication *app) {
  // Create a new window
  GtkWidget *window = gtk_application_window_new (app);
  // Create a new button
  GtkWidget *button = gtk_button_new_with_label ("Hello, World!");
  // When the button is clicked, close the window passed as an argument
  g_signal_connect_swapped (button, "clicked", G_CALLBACK (gtk_window_close), window);
  gtk_window_set_child (GTK_WINDOW (window), button);
  gtk_window_present (GTK_WINDOW (window));
}

int main (int argc, char *argv[]) {
  // Create a new application
  GtkApplication *app = gtk_application_new ("com.example.GtkApplication",
                                             G_APPLICATION_FLAGS_NONE);
  g_signal_connect (app, "activate", G_CALLBACK (on_activate), NULL);
  return g_application_run (G_APPLICATION (app), argc, argv);
}
```


I went to project properties > Linker > System and set SubSystem to Console (/SUBSYSTEM:CONSOLE)

I attempted to run build the project and got into the issue reported in this ticket.

To fix it I went to Project Properties > C/C++ > General > Additional Include Directories 
```
$(VcpkgRoot)installed\x86-windows\include\gtk-4.0;$(VcpkgRoot)installed\x86-windows\include\pango-1.0;$(VcpkgRoot)installed\x86-windows\include\cairo;$(VcpkgRoot)installed\x86-windows\include\gio-win32-2.0;$(VcpkgRoot)installed\x86-windows\include\glib-2.0;$(VcpkgRoot)installed\x86-windows\\lib\glib-2.0\include;$(VcpkgRoot)installed\x86-windows\include\\gdk-pixbuf-2.0;$(VcpkgRoot)installed\x86-windows\include\graphene-1.0;$(VcpkgRoot)installed\x86-windows\include\harfbuzz;$(VcpkgRoot)installed\x86-windows\lib\graphene-1.0\include;$(VcpkgRoot)installed\x86-windows\tools\glib
```
I had to build this path manually based on the errors from the build step.
Not sure how else to generate it. 
I read something about `pkg-config.exe` but when I tried to execute it doesn't find the packages from vcpkg.
```
$ pkg-config.exe --cflags gtk-4.0
Package gtk-4.0 was not found in the pkg-config search path.
Perhaps you should add the directory containing `gtk-4.0.pc'
to the PKG_CONFIG_PATH environment variable
No package 'gtk-4.0' found
```
Maybe vcpkg can somehow generate a value for that `PKG_CONFIG_PATH` environment variable so they can work together ?

After modifying the Additional Include Directories the build finally succeeded.

When i click on the start debugging button in visual studio I get this error
![image](https://user-images.githubusercontent.com/3531898/126132621-9a139792-3693-4475-a473-c8688079c56e.png)
```
Exception thrown at 0x759583DC (kernel32.dll) in ExampleGTKProject.exe: 0xC0000005: Access violation writing location 0x0000000B.

The program '[22812] ExampleGTKProject.exe' has exited with code 0 (0x0).
```

Not sure what I can do about it. 

In the console that opens I get the following warning:
```
(ExampleGTKProject.exe:22812): GLib-GIO-WARNING **: 11:56:01.802: win32 session dbus binary not found: C:\Users\ionutvmi\source\repos\ExampleGTKProject\Debug\\gdbus.exe
```

When I go to the debug folder I manually run the generated exe it works:
![image](https://user-images.githubusercontent.com/3531898/126133130-93a2201c-05e1-4b52-bc01-3a28b821bded.png)
I still get the gdbus warning in the console. 
I guess I have to manually copy that exe from `$(VcpkgRoot)installed\x86-windows\tools\glib` to the Debug folder ??

Are there other GUI libraries that work better in this vcpkg + visual studio integration ?


The overall moral of the story is that next weekend I will be watching netflix instead.

**Comment by dermuedejones** [comment]
i am facing the same issue with the gtk4 integration. With gtk it was really easy as there was no nested directory structure for the version. Is is planned to have multiple versions of the same library installed for a vcpkg root? With the new option to specify an install dir I think it is not necessary, and the includes should be provided as before.

I am glad you opened this thread as I thought somewhat broke with my integration during the update, but this confirms that the directories have changed.

**Comment by PJBoy** [comment]
Any plans for this issue? If it ends up being a "won't fix", then I'll be raising a bug on the readme file on the snippet quoted by Hemaolle

> Yeah, at least the vcpkg readme promises:
> 
> > After this, you can now create a New non-CMake Project (or open an existing one). All installed libraries are immediately ready to be #include'd and used in your project without additional configuration.

**Comment by JackBoosY** [comment]
If the upstream behavior causes the header file to be installed in _gtk-4.0_, this is by design and conflicts with the example code provided by the upstream.
If this is caused by vcpkg, we need to fix this issue.

**Comment by dg0yt** [comment]
> If the upstream behavior causes the header file to be installed in _gtk-4.0_, this is by design and conflicts with the example code provided by the upstream.
> If this is caused by vcpkg, we need to fix this issue.

IIUC the problem is (convenience in) Visual Studio:
How to add `include/gtk-4.0` to the search path when not using cmake or pkgconfig?
This is what needs to be fixed by vcpkg when vcpkg promises "you can now create a New non-CMake Project (or open an existing one). All installed libraries are immediately ready to be #include'd and used in your project without additional configuration."

**Reply by sullivak-kyra** [reply]
I'm running into the same issue with itk. I'm using a manifest but issue is the same, I need to manually add a "..../include/ITK-5.1" to additional include directories to use itk and nested (e.g. vnl) header files.
